server/src/main/java/org/elasticsearch/common/time/DateUtils.java

```diff
@@ -293,6 +293,17 @@ public static long toMilliSeconds(long nanoSecondsSinceEpoch) {
         return nanoSecondsSinceEpoch / 1_000_000;
     }
 
+    public static int compareNanosToMillis(long nanos, long millis) {
+        if (millis < 0) {
+            return 1;
+        }
+        if (millis > MAX_NANOSECOND_IN_MILLIS) {
```

Comment:
This seems like it should be a precondition. Why would we have millis outside the expected input bounds?

Reply:
The intent here is to support comparing user data stored in the two different resolutions. So in this case, the millis aren't outside the expected input bounds, as any long-representable datetime should be a valid input.  A user might have a query like `WHERE nanos_date < millis_date`, and `millis_date` could be past the maximum long-representable nanosecond date, in which case we would still want that expression to evaluate to `true`.  We can optimize that away if the milliseconds are a constant, but if they're in a field, we actually need to compare and get a sensible result.